runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -8383,28 +8383,41 @@ class INTERPRETER_CLASS
 		J9Method *method = (J9Method *)(UDATA)J9OBJECT_U64_LOAD(_currentThread, memberNameObject, _vm->vmtargetOffset);
 		UDATA vTableOffset = methodID->vTableIndex;
 
-		if (J9_ARE_ANY_BITS_SET(vTableOffset, J9_JNI_MID_INTERFACE)) {
-			UDATA iTableIndex = vTableOffset & ~(UDATA)J9_JNI_MID_INTERFACE;
-			J9Class *interfaceClass = J9_CLASS_FROM_METHOD(method);
-			vTableOffset = 0;
-			J9ITable * iTable = receiverClass->lastITable;
-			if (interfaceClass == iTable->interfaceClass) {
-				goto foundITable;
-			}
-			iTable = (J9ITable*)receiverClass->iTable;
-			while (NULL != iTable) {
+		/* vmindexOffset (J9JNIMethodID) is initialized using jnicsup.cpp::initializeMethodID.
+		 * initializeMethodID will set J9JNIMethodID->vTableIndex to 0 for private interface
+		 * methods and j.l.Object methods. When J9JNIMethodID->vTableIndex is 0, then
+		 * vmtargetOffset (J9Method) is the _sendMethod, and it will point to the private
+		 * interface method or j.l.Object method. When J9JNIMethodID->vTableIndex is not 0,
+		 * then it is either a vTable offset or an iTable index.
+		 */
+		if (0 == vTableOffset) {
+			/* Private interface method or j.l.Object method. */
+			_sendMethod = method;
+		} else {
+			/* Treat as vTable offset for the method if J9_JNI_MID_INTERFACE is not set. */
+			if (J9_ARE_ANY_BITS_SET(vTableOffset, J9_JNI_MID_INTERFACE)) {
+				/* Treat as iTable index for the method if J9_JNI_MID_INTERFACE is set. */
+				UDATA iTableIndex = vTableOffset & ~(UDATA)J9_JNI_MID_INTERFACE;
+				J9Class *interfaceClass = J9_CLASS_FROM_METHOD(method);
```

Comment:
The `MemberName->vmtargetOffset` is directly holding onto a `J9Method*`.  How is that pointer updated during HCR?  I'm concerned that the loop through the `iTable` may fail to match the `interfaceClass` if HCR has replaced the class with a newer version.  (ie:  The cached J9Method * may hold onto an obsolete class pointer)

Reply:
Should `VM_VMHelpers::currentClass` get the most current version of class after HCR?
```suggestion
				J9Class *interfaceClass = VM_VMHelpers::currentClass(J9_CLASS_FROM_METHOD(method));
```

Do we also need to update the `J9Method` ref stored in `MemberName` for HCR? Related code:
- [hshelp.c::fixDirectHandles](https://github.com/eclipse/openj9/blob/b58ed1b52baf49092e3ac19cdcf8058b17b777db/runtime/util/hshelp.c#L2092-L2141): Here `J9Methods` stored in OpenJ9 MHs are updated. We probably need to do the same for `MemberNames->vmtarget` (`J9Method`).
- [hshelp.c::fixJNIRefs](https://github.com/eclipse/openj9/blob/b58ed1b52baf49092e3ac19cdcf8058b17b777db/runtime/util/hshelp.c#L1676-L1701) will need to be invoked before fixing `MemberNames->vmindex` (`J9JNIMethodID`). 
- [jvmtiClass.c::redefineClassesCommon](https://github.com/eclipse/openj9/blob/b58ed1b52baf49092e3ac19cdcf8058b17b777db/runtime/jvmti/jvmtiClass.c#L1088): invokes `hshelp.c::fixJNIRefs` and `hshelp.c::fixDirectHandles`.
- Related question: https://github.com/eclipse/openj9/pull/10690#discussion_r531728225

Reply:
> Should `VM_VMHelpers::currentClass` get the most current version of class after HCR?

This solves half the problem in that it ensures the most up-to-date version of the class is used in the itable search.  Unfortunately, the itableIndex may not point to the same method in the new version of the class if the methods have been reordered.

It's really two pieces of data - class version & itable index - and both need to kept consistent and up to date.

> * [hshelp.c::fixDirectHandles](https://github.com/eclipse/openj9/blob/b58ed1b52baf49092e3ac19cdcf8058b17b777db/runtime/util/hshelp.c#L2092-L2141): Here `J9Methods` stored in OpenJ9 MHs are updated. We probably need to do the same for `MemberNames->vmtarget` (`J9Method`).

This is carefully set up thru cooperation between the MH creation java code & the redef code to avoid a full heap walk to find the DirectHandles.  We may be able to make the same invariants hold for MemberName but I'm less clear on how to do that without patching the OJDK java code.

Couple of options here:
1) Do a full heap walk on redefinition to fix the `vmtarget` & `vmindex` fields of the MemberName.  This will regress the use cases we fixed by adding the DirectHandle cache but it will be correct.

2) Change `vmindex` to be an index into class->jniIDs table.  This should let the existing mechanisms fix the JNIID when the class is redefined but will require changes to the way invocation occurs.  We'll also need to figure out how to update `vmtarget` as well.

3) Something else?

I'd suggest starting with option 1 as it's the easiest to get right and we can figure out how to optimize it later.

Reply:
I have updated this PR to handle the `VM_VMHelpers::currentClass` change. For the bigger part, I have opened the following issue: https://github.com/eclipse/openj9/issues/11528. This PR is ready to be reviewed and merged. The bigger part will be handled in a separate PR.